### ISSUE TITLE
Repository iterators don't get closed and don't release resources until the repository is closed.

### ISSUE THREAD

**Issue by fdlk** [molgenis/molgenis#3499]
When retrieving Entities from a `Repository`, we usually call the `iterator` or `query` methods which return an `Iterator<Entity>`.

If the Repository is backed by a file, it'll usually create some kind of input stream that needs to be closed when the iterator is done.
There is no general way for the Repository to tell that the iterator is no longer being used. It can detect when the last element gets read, but perhaps the client is only interested in the first couple Entities.
There is no way for the client to notify the Repository that it's done reading from the iterator.

As a result, the TabixVcfRepository leaves all the VcfReaders it creates (one per annotated row if it is being used as an annotation resource) open until the repository itself is closed. Which is way too late.

**Comment by dennishendriksen** [comment]
org.molgenis.util.CloseableIterator might be helpful here.

**Comment by fdlk** [comment]
Yes, that would help. Or possibly java.util.Stream<Entity> which is AutoCloseable and has an onClose method where you can add a handler.

**Comment by Mark-de-Haan** [comment]
@fdlk Does this still happen with the new stream implementations?

**Comment by dennishendriksen** [comment]
Relevant https://github.com/molgenis/molgenis/issues/5154
